pkg/cloud/services/iamauth/reconcile.go

```diff
@@ -21,23 +21,17 @@ import (
 	"fmt"
 
 	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/aws/aws-sdk-go/service/iam"
 	"github.com/pkg/errors"
 	"k8s.io/klog/v2"
 
 	ekscontrolplanev1 "sigs.k8s.io/cluster-api-provider-aws/v2/controlplane/eks/api/v1beta2"
-	iamv1 "sigs.k8s.io/cluster-api-provider-aws/v2/iam/api/v1beta1"
 )
 
 // ReconcileIAMAuthenticator is used to create the aws-iam-authenticator in a cluster.
-func (s *Service) ReconcileIAMAuthenticator(ctx context.Context) error {
+func (s *Service) ReconcileIAMAuthenticator(ctx context.Context, nodeRoles map[string]struct{}) error {
```

Comment:
I see that this function gets other roles from `s.scope` . Could it fetch `nodeRoles`  in a similar way? Would that be an improvement over making `nodeRoles` a function argument?

Reply:
This is a good question, and i'm willing to defer to whatever choice maintainers want to make. 

To me, the call to get the node roles in the aws managed control plane reconciler and passing on to this  `ReconcileIAMAuthenticator` read clearly. 

In the scope it could be a little harder to figure out.

Reply:
·The pattern I see is that "services" are [initialized with a scope](https://github.com/kubernetes-sigs/cluster-api-provider-aws//blob/800603e6bae2a049fdf9de0f7bae3045be78885f/controlplane/eks/controllers/awsmanagedcontrolplane_controller.go#L251-L257), and define "reconcile" methods with only a `context.Context` argument, e.g., https://github.com/kubernetes-sigs/cluster-api-provider-aws//blob/800603e6bae2a049fdf9de0f7bae3045be78885f/controlplane/eks/controllers/awsmanagedcontrolplane_controller.go#L259.

Reusing this pattern could help long-term maintenance. I'll explore a refactoring and update here.

Reply:
Thanks for refactoring